README.md
# Hugo Invi - The Fresh theme for Hugo

**Hugo Invi** is a multi-page website theme for the [Hugo](https://gohugo.io) static site generator, based on [hugo-fresh](https://github.com/StefMa/hugo-fresh) but heavily modified. hugo-fresh itself is adapted from [CSS Ninja](https://cssninja.io/product/fresh)'s `Fresh`, a bulma-based landing page.

## Functionality

Hugo Invi supports the following features…

Technical
* Internationalization
  * incl. an option in the Navigation Bar to change the site language
* Meta Data for SEO; supports
  * OpenGraph
  * Twitter
  * Schema
    * Organization
* Color Palette
* various Shortcodes and Markdown Render Hooks

Visual
* Navigation Bar
* Sidebar
* Image Gallery
* Embedding YouTube videos
  * incl. thumbnail with warning as video replacement: user has to accept YouTube cookies first (decision is saved as a cookie)
* Feature / Product cards
  * varying height or aligned height
* Short list of main features (large font size)
* Quote Cards
* Cookie Notification

## Usage

### Custom Javascript

* Write custom javascript and declare the path in config file using `params.customJS`
  * layouts/partials/javascript.html will fetch and insert it in each page including this partial (as of now: `baseof.html`, `single.html`)


### Merge Project styling with this theme

* Write custom sass file in your `assets` folder and declare the path in config file using `params.customCSS`
  * The styling will be merged with this theme


### Content Shortcodes

[Hugo Docs](https://gohugo.io/content-management/shortcodes/)

Title and Subtitle
```css
// id is optional
{{< title id="delicious" lvl="3" text="Bacon" >}}

// id is optional
{{< subtitle id="delicious" lvl="3" text="Bacon" >}}
```

### Markdown Render Hooks

[Hugo Docs](https://gohugo.io/templates/render-hooks/)

supported
* render link
  * if the URL has the "http" prefix, it gains `target="_blank" rel="noopener, nofollow"`

CHANGELOG.md
# Changelog

All notable changes to this project (since its fork) are documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [2.0.2] - 2023-02-06

### Added

* config option: `footer.useColumns`
  * if set to false, toggle footer style to left-aligned (instead of columns).


### Changed

* exampleSite/config.yaml > `baseURL` changed to localhost:1313.
* If content file for hero-item.html provides no button text, hide button.
* If content file for hero.html provides no image, just show header horizontally centered (and hide button if no text defined).


## [2.0.1] - 2023-01-31

### Added

* Support including website's custom javascript (see README.md)
* Support merging website's custom styling into theme's styling (see README.md)


### Changed

* Moved `static/images` to subfolder exampleSite.


### Removed

* `images/` folder.

 
### Fixed

* Placement of cookie notification: now truly stays at the window's bottom.
* Defined `sameSite` value for cookies (web browser was showing warnings).


## [2.0] - 2023-01-20

### Added

* Localization support
  * in config.yaml: navbar, sidebar, footer.
  * in landing page.
* SEO meta support for
  * OpenGraph
  * Twitter
  * Schema
    * Organization
* Other SEO
  * page param "seo_title"
* Default Archetype
* Image Gallery support
* YouTube Video support
* Website Cookies notification
* Color Palette support to recolor this theme


### Changed

* Landing page site data (previously in config.yaml) is now defined in website's index.md file's front matter (i.e. SITE-variables were changed to PAGE-varibles).
* All URLs are expected to be internal and respect the user's currently selected language; except for cases where external URLs are expected, e.g. for social media.
* Property names are called "name" instead of "title", if they represent e.g. menu options in the navbar exactly that; and not "link", which is the element


### Removed

* Site param `include_footer` from front matter; show footer at all times.
* Flex Card styling (is unused)


### Fixed

* Property names are called "url", if they are exactly that; and not "link", which is the element.
* Accessed wrong site param in navbar-clone for navbar logo: should be height, not width


### Security

* Add `noopener, nofollow` for offsite URLs.


## [1.1] - 2022-12-10

### Added

* Support for copyright notice
* "_markup/render-link": adds *noopener* and *nofollow* relation to markdown links in `content/`, if external link (i.e. URL begins with "http")
* Allow configuring Feature Card height


### Changed

* Condense subtitle{number}.html shortcodes
  * Breaking Change: see exampleSite/content/agb.md
* Condense title{number}.html shortcodes
  * Breaking Change: see exampleSite/content/agb.md


## [1.0] - 2022-12-08

> Basically a cleanup of hugo-fresh while preserving functionality


### Added

* exampleSite: added gohugo privacy settings to config.yml


### Changed

* Upgrade fontawesome to 6.2.1-web
* Upgrade bulma to 0.9.4
* Update jQuery to 3.6.1
* Default page background color is dark (defines footer color), sections overwrite to white/grey
* Replace hard-coded section color with "every nth-child"-logic


### Removed

* jquery.panelslider (unused)
* feather-icons (replaced by fontawesome icons)
* google font "Open Sans" (also removes google api)


### Fixed

* Images in sections not shown in some browsers (previously, it was only working for card icons)
* If javascript was disabled, preloader animation was playing indefinitely


static/js/fresh.js
//Preloader
$(window).on('load', function () { // makes sure the whole site is loaded 
    $('#status').fadeOut(); // will first fade out the loading animation 
    $('#preloader').delay(350).fadeOut('slow'); // will fade out the white DIV that covers the website. 
    $('body').delay(350).css({ 'overflow': 'visible' });
})

$(document).ready(function () {
    //Mobile menu toggle
    if ($('.navbar-burger').length) {
        $('.navbar-burger').on("click", function () {

            var menu_id = $(this).attr('data-target');
            $(this).toggleClass('is-active');
            $("#" + menu_id).toggleClass('is-active');
            $('.navbar.is-light').toggleClass('is-dark-mobile')
        });
    }

    //Animate left hamburger icon and open sidebar
    $('.menu-icon-trigger').on("click", function (e) {
        e.preventDefault();
        $('.menu-icon-wrapper').toggleClass('open');
        $('.sidebar').toggleClass('is-active');
    });

    //Close sidebar
    $('.sidebar-close').on("click", function () {
        $('.sidebar').removeClass('is-active');
        $('.menu-icon-wrapper').removeClass('open');
    })

    //Sidebar menu
    if ($('.sidebar').length) {
        $(".sidebar-menu > li.have-children > a").on("click", function (i) {
            i.preventDefault();
            if (!$(this).parent().hasClass("active")) {
                $(".sidebar-menu li ul").slideUp();
                $(this).next().slideToggle();
                $(".sidebar-menu li").removeClass("active");
                $(this).parent().addClass("active");
            }
            else {
                $(this).next().slideToggle();
                $(".sidebar-menu li").removeClass("active");
            }
        });
    }

    //Navbar Clone
    if ($('#navbar-clone').length) {
        $(window).on("scroll", function () {    // this will work when your window scrolled.
            var height = $(window).scrollTop();  //getting the scrolling height of window
            if (height > 50) {
                $("#navbar-clone").addClass('is-active');
            } else {
                $("#navbar-clone").removeClass('is-active');
            }
        });
    }

    //reveal elements on scroll so animations trigger the right way
    var $window = $(window),
        win_height_padded = $window.height() * 1.1,
        isTouch = Modernizr.touch;

    $window.on('scroll', revealOnScroll);

    function revealOnScroll() {
        var scrolled = $window.scrollTop();
        $(".revealOnScroll:not(.animated)").each(function () {
            var $this = $(this),
                offsetTop = $this.offset().top;

            if (scrolled + win_height_padded > offsetTop) {
                if ($this.data('timeout')) {
                    window.setTimeout(function () {
                        $this.addClass('animated ' + $this.data('animation'));
                    }, parseInt($this.data('timeout'), 10));
                } else {
                    $this.addClass('animated ' + $this.data('animation'));
                }
            }
        });
    }

    // Back to Top button behaviour
    var pxShow = 600;
    var scrollSpeed = 500;
    $(window).on("scroll", function () {
        if ($(window).scrollTop() >= pxShow) {
            $("#backtotop").addClass('visible');
        } else {
            $("#backtotop").removeClass('visible');
        }
    });
    $('#backtotop a').on('click', function () {
        $('html, body').animate({
            scrollTop: 0
        }, scrollSpeed);
        return false;
    });

    // Select all links with hashes
    $('a[href*="#"]')
        // Remove links that don't actually link to anything
        .not('[href="#"]')
        .not('[href="#0"]')
        .on("click", function (event) {
            // On-page links
            if (
                location.pathname.replace(/^\//, '') == this.pathname.replace(/^\//, '')
                &&
                location.hostname == this.hostname
            ) {
                // Figure out element to scroll to
                var target = $(this.hash);
                target = target.length ? target : $('[name=' + this.hash.slice(1) + ']');
                // Does a scroll target exist?
                if (target.length) {
                    // Only prevent default if animation is actually gonna happen
                    event.preventDefault();
                    $('html, body').animate({
                        scrollTop: target.offset().top
                    }, 550, function () {
                        // Callback after animation
                        // Must change focus!
                        var $target = $(target);
                        $target.focus();
                        if ($target.is(":focus")) { // Checking if the target was focused
                            return false;
                        } else {
                            $target.attr('tabindex', '-1'); // Adding tabindex for elements not focusable
                            $target.focus(); // Set focus again
                        };
                    });
                }
            }
        });
})


static/js/invi.js
import Cookies from '../modules/js.cookie.esm.min.js';
import PhotoSwipeLightbox from '../modules/photoswipe/photoswipe-lightbox.esm.min.js'; // also imports core
import PhotoSwipe from '../modules/photoswipe/photoswipe.esm.min.js';

//Preloader
$(window).on('load', function () { // makes sure the whole site is loaded 
    $('#status').fadeOut(); // will first fade out the loading animation 
    $('#preloader').delay(350).fadeOut('slow'); // will fade out the white DIV that covers the website. 
    $('body').delay(350).css({ 'overflow': 'visible' });
})

$(document).ready(function () {
    // Handle Website Cookies
    {
        var notification = $('#cookie-notification');
        if (notification.length) {
            var allow_other_cookies = Cookies.get("allow-nonfunctional-website-cookies");
            if (allow_other_cookies !== null && typeof allow_other_cookies !== "undefined") {
                notification.hide();
            }
        }

        if ($('#decline-cookies').length) {
            $('#decline-cookies').on("click", function () {
                handleWebsiteCookies(false);
            });
        }
        if ($('#accept-cookies').length) {
            $('#accept-cookies').on("click", function () {
                handleWebsiteCookies(true);
            });
        }
    }

    // Initialize image gallery
    const lightbox = new PhotoSwipeLightbox({
        gallery: '#invi-gallery',
        children: 'a',
        pswpModule: PhotoSwipe,
    });
    lightbox.init();

    // Handle video embedding
    {
        // Check on page load if user gave consent to youtube cookies.
        // If true, show video immediately.
        // If false, show placeholder image and request for consent or direct link.
        var allow_youtube_cookies = Cookies.get("allow-youtube-cookies");

        var notification = document.getElementsByClassName("video-embed-notification")[0];
        if (notification !== null && typeof notification !== "undefined") {
            if (allow_youtube_cookies === null) {
                notification.style.display = "block";
            } else {
                if (allow_youtube_cookies === "true") {
                    notification.style.display = "none";
                } else {
                    notification.style.display = "block";
                }
            }
        }

        if (allow_youtube_cookies === "true") {
            var embed_video = document.getElementsByClassName("embed-video")[0];
            if (typeof embed_video === "undefined") {
                // was undefined, append video
                var video_id = getVideoID();
                appendVideo(video_id);
            }
        }

        if ($('.show-video').length) {
            $('.show-video').on("click", acceptCookiesAndShowYouTubeVideo);
        }
    }

    // Mobile menu toggle
    if ($('.navbar-burger').length) {
        $('.navbar-burger').on("click", function () {
            var menu_id = $(this).attr('data-target');
            $(this).toggleClass('is-active');
            $("#" + menu_id).toggleClass('is-active');
            $('.navbar.is-light').toggleClass('is-dark-mobile')
        });
    }

    // Animate left hamburger icon and open sidebar
    $('.menu-icon-trigger').on("click", function (e) {
        e.preventDefault();
        $('.menu-icon-wrapper').toggleClass('open');
        $('.sidebar').toggleClass('is-active');
    });

    // Close sidebar
    $('.sidebar-close').on("click", function () {
        $('.sidebar').removeClass('is-active');
        $('.menu-icon-wrapper').removeClass('open');
    })

    // Sidebar menu
    if ($('.sidebar').length) {
        $(".sidebar-menu > li.have-children > a").on("click", function (i) {
            i.preventDefault();
            if (!$(this).parent().hasClass("active")) {
                $(".sidebar-menu li ul").slideUp();
                $(this).next().slideToggle();
                $(".sidebar-menu li").removeClass("active");
                $(this).parent().addClass("active");
            }
            else {
                $(this).next().slideToggle();
                $(".sidebar-menu li").removeClass("active");
            }
        });
    }

    // Navbar Clone
    if ($('#navbar-clone').length) {
        $(window).on("scroll", function () {    // this will work when your window scrolled.
            var height = $(window).scrollTop(); // getting the scrolling height of window
            if (height > 50) {
                $("#navbar-clone").addClass('is-active');
            } else {
                $("#navbar-clone").removeClass('is-active');
            }
        });
    }

    // Reveal elements on scroll so animations trigger the right way
    var $window = $(window),
        win_height_padded = $window.height() * 1.1,
        isTouch = Modernizr.touch;

    $window.on('scroll', revealOnScroll);

    function revealOnScroll() {
        var scrolled = $window.scrollTop();
        $(".revealOnScroll:not(.animated)").each(function () {
            var $this = $(this),
                offsetTop = $this.offset().top;

            if (scrolled + win_height_padded > offsetTop) {
                if ($this.data('timeout')) {
                    window.setTimeout(function () {
                        $this.addClass('animated ' + $this.data('animation'));
                    }, parseInt($this.data('timeout'), 10));
                } else {
                    $this.addClass('animated ' + $this.data('animation'));
                }
            }
        });
    }

    // Back to Top button behaviour
    var pxShow = 600;
    var scrollSpeed = 500;
    $(window).on("scroll", function () {
        if ($(window).scrollTop() >= pxShow) {
            $("#backtotop").addClass('visible');
        } else {
            $("#backtotop").removeClass('visible');
        }
    });
    $('#backtotop a').on('click', function () {
        $('html, body').animate({
            scrollTop: 0
        }, scrollSpeed);
        return false;
    });

    // Select all links with hashes
    $('a[href*="#"]')
        // Remove links that don't actually link to anything
        .not('[href="#"]')
        .not('[href="#0"]')
        .on("click", function (event) {
            // On-page links
            if (
                location.pathname.replace(/^\//, '') == this.pathname.replace(/^\//, '')
                &&
                location.hostname == this.hostname
            ) {
                // Figure out element to scroll to
                var target = $(this.hash);
                target = target.length ? target : $('[name=' + this.hash.slice(1) + ']');
                // Does a scroll target exist?
                if (target.length) {
                    // Only prevent default if animation is actually gonna happen
                    event.preventDefault();
                    $('html, body').animate({
                        scrollTop: target.offset().top
                    }, 550, function () {
                        // Callback after animation
                        // Must change focus!
                        var $target = $(target);
                        $target.focus();
                        if ($target.is(":focus")) { // Checking if the target was focused
                            return false;
                        } else {
                            $target.attr('tabindex', '-1'); // Adding tabindex for elements not focusable
                            $target.focus(); // Set focus again
                        };
                    });
                }
            }
        });
})

/**
 * handleWebsiteCookies saves user preferences regarding saving cookies.
 * @return {boolean} whether to accept cookies beyond the necessary functional ones.
 */
function handleWebsiteCookies(accept) {
    Cookies.set("allow-nonfunctional-website-cookies", accept, { expires: 730, sameSite: 'lax' }); // i.e. in 2 years

    var notification = $("#cookie-notification");
    if (notification.length) {
        notification.hide();
    }
}

/**
 * acceptCookiesAndShowYouTubeVideo user clicked on youtube cookie consent button beneath video placeholder.
 */
function acceptCookiesAndShowYouTubeVideo() {
    Cookies.set("allow-youtube-cookies", true, { expires: 730, sameSite: 'lax' }); // i.e. in 2 years

    // hide consent notification
    var notification = document.getElementsByClassName("video-embed-notification")[0];
    if (notification !== null) {
        notification.style.display = "none";
    }

    var video_id = getVideoID();
    appendVideo(video_id);
}

/**
 * getVideoID returns the video id or an empty string.
 * @return {string} The video id to load
 */
function getVideoID() {
    // get video id
    var consent_button = document.getElementsByClassName("show-video")[0];
    if (consent_button === null) {
        return "";
    }
    var video_id = consent_button.getAttribute("data-video-id")
    if (video_id === null) {
        return "";
    }

    return video_id;
}

/**
 * appendVideo inserts iframe requires to display the video.
 * @param {string} video_id The video id to load
 */
function appendVideo(video_id) {
    var wrapper = document.getElementsByClassName("embedded-video-parent")[0];
    wrapper.style.display = "block";
    const span = document.createElement("span");
    span.innerHTML =
        '<iframe class="embed-video" allowfullscreen="" src="https://www.youtube-nocookie.com/embed/' + video_id + '?rel=0&iv_load_policy=3" frameborder="0"></iframe>';
    wrapper.appendChild(span);
}
